### ISSUE TITLE
branch-governance and feature-branch-protection disagree on valid branch names; chore/ is suggested by one and rejected by the other

### ISSUE THREAD

**Issue by ruleofcodedev** [Rule-Of-Code/RoC#80]
## Version

`ruleofcode@7.17.6`

## Law

Two laws, one concept, two different rules:

- `branch-governance` — `BranchGovernanceLaw` in `dist/checkers/git-laws/branch-governance.js`
- `feature-branch-protection` — `FeatureBranchProtectionLaw` in `dist/checkers/git-laws/feature-branch-protection.js`

## What's wrong

Both validate the current branch name, independently, with rules that do not agree.

`branch-governance` uses an anchored regex:

```js
const validBranchPattern = /^(feature|hotfix|bugfix|release)\/[a-z0-9.-]+$/;
...
else if (!validBranchPattern.test(currentBranch)) {
    violations.push(`Branch name "${currentBranch}" doesn't follow convention: feature/hotfix/bugfix/release/<name>`);
}
```

`feature-branch-protection` uses a prefix list:

```js
const validPrefixes = [
    'feature/',
    'bugfix/',
    'hotfix/',
    'release/',
    'chore/',
];
const followsConvention = validPrefixes.some(prefix => currentBranch.startsWith(prefix));
if (!followsConvention) {
    violations.push(`Branch '${currentBranch}' doesn't follow naming convention`);
    suggestions.push('Use branch prefixes: feature/, bugfix/, hotfix/, release/, chore/');
}
```

The consequences, checked by running both rules against the same inputs rather than reasoning about them:

| Branch | `branch-governance` | `feature-branch-protection` |
|---|---|---|
| `feature/add-user-auth` | PASS | PASS |
| `chore/update-deps` | **FAIL** | PASS |
| `feature/Add_User_Auth` | **FAIL** | PASS |
| `feature/roc/sub-scope` | **FAIL** | PASS |
| `fix/some-thing` | FAIL | FAIL |

Three distinct disagreements, each from a different cause:

1. **`chore/` is recommended by one law and rejected by the other.** This is the sharpest one: `feature-branch-protection` lists `chore/` in `validPrefixes` *and* prints `Use branch prefixes: feature/, bugfix/, hotfix/, release/, chore/` as its remediation advice. A user who follows that advice gets a `branch-governance` violation, because `chore` is absent from its alternation. RoC's own suggestion produces a RoC failure.
2. **Character class.** `[a-z0-9.-]+` rejects uppercase and underscores; the prefix check has no charset opinion at all, so `feature/Add_User_Auth` passes one and fails the other.
3. **Nesting.** `/` is not in `[a-z0-9.-]`, so a nested name like `feature/roc/sub-scope` fails `branch-governance` while passing `feature-branch-protection`.

Note the last row: `fix/` fails both. That is consistent behaviour and not what this report is about — the two laws agreeing on a policy is fine, even if the policy is stricter than conventional-commits users expect.

## Reproduction (our repo)

Naming a branch `fix/devops-release-image-tag-namespaced` produced two violations for one cause:

```
❌ FAILED CONSTITUTIONAL LAWS:
1. Branch Governance Standards [1 violations] (Score: 90/100)
   • Branch name "fix/devops-release-image-tag-namespaced" doesn't follow convention: feature/hotfix/bugfix/release/<name>
2. Feature Branch Protection [1 violations] (Score: 90/100)
   • Branch 'fix/devops-release-image-tag-namespaced' doesn't follow naming convention
```

Renaming to `bugfix/…` cleared both. That part is working as intended; investigating *why* two laws fired for one name is what surfaced the divergence above. The table was produced by extracting both rules verbatim from `dist/` and running them over the same list.

## Suggested fix

- Share one branch-name validator between the two laws, the way this codebase already shares `GitLawBase` — the drift here is the same sibling-implementation pattern as [#61](https://github.com/Rule-Of-Code/RoC/issues/61)/[#62](https://github.com/Rule-Of-Code/RoC/issues/62) and [#76](https://github.com/Rule-Of-Code/RoC/issues/76)/[#77](https://github.com/Rule-Of-Code/RoC/issues/77).
- Decide `chore/` deliberately in that one place. Whichever way it goes, the suggestion string and the accepted set must not contradict each other.
- Consider allowing `/` inside the name portion (nested scopes) and deciding uppercase/underscore explicitly rather than by an incidental character class.
- If the two laws are meant to check genuinely different things, it would help to say so in the violation messages — as written, both read as "your branch name is wrong", which is why the divergence is invisible until the rules are compared side by side.

**Comment by ruleofcodedev** [comment]
Fixed in #87, on `develop`.

Both laws ask one validator now. Your sharpest point — that we recommended `chore/` and then failed it — is fixed at the root: **the suggestion string is generated from the accepted set**, so the advice and the policy cannot contradict each other again.

Confirmation that this was not theoretical: it bit us the same day you filed it. A back-merge branch named `chore/backmerge-v7.19.0` failed our own CI on Branch Governance while Feature Branch Protection was content with it.

Your table, re-run after the fix, plus a fourth row the sweep added:

| Branch | before | after |
| --- | --- | --- |
| `chore/update-deps` | governance rejected | both accept |
| `feature/Add_User_Auth` | `_` rejected | both accept |
| `feature/roc/sub-scope` | `/` rejected | both accept |
| `feature/ABC-123` | **not reported by you** — uppercase rejected | both accept |
| `fix/some-thing` | both rejected | both reject |

That fourth row is the point you made about the character class deciding policy by accident. Nobody chose to reject a branch named after a ticket; `[a-z0-9.-]+` did. The accepted set is explicit now, and configurable via `thresholds.git.branchNaming.allowedPrefixes` — `fix/` is still rejected by default, which as you noted is a policy question rather than a bug, and it is now a policy you can change in one place.

On your last suggestion: the two laws keep separate messages, because once the name is valid they check different things — governance is about the name, protection is about what may be pushed where. Reading both as "your branch name is wrong" was a symptom of the shared rule being duplicated, not of the messages.
